File: pki/base/common/src/com/netscape/cmscore/usrgrp/Group.java
// --- BEGIN COPYRIGHT BLOCK ---
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; version 2 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, write to the Free Software Foundation, Inc.,
// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
//
// (C) 2007 Red Hat, Inc.
// All rights reserved.
// --- END COPYRIGHT BLOCK ---
package com.netscape.cmscore.usrgrp;


import java.util.*;
import com.netscape.certsrv.base.*;
import com.netscape.certsrv.usrgrp.*;
import com.netscape.certsrv.apps.*;


/**
 * A class represents a group.
 *
 * @author cfu
 * @version $Revision: 14561 $, $Date: 2007-05-01 10:28:56 -0700 (Tue, 01 May 2007) $
 */
public class Group implements IGroup {
    private IUsrGrp mBase = null;
    private String mName = null;
    private Vector mMembers = new Vector();
    private String mDescription = null;

    private static final Vector mNames = new Vector();
    static {
        mNames.addElement(ATTR_NAME);
        mNames.addElement(ATTR_ID);
        mNames.addElement(ATTR_DESCRIPTION);
        mNames.addElement(ATTR_MEMBERS);
    }

    /**
     * Constructs local group.
     */
    public Group(IUsrGrp base, String name) {
        mBase = base;
        mName = name;
    }

    public String getName() {
        return mName;
    }

    public String getGroupID() {
        return mName;
    }

    public String getDescription() {
        return mDescription;
    }

    public void addMemberName(String name) {
        mMembers.addElement(name);
    }

    public Enumeration getMemberNames() {
        return mMembers.elements();
    }

    public boolean isMember(String name) {
        for (int i = 0; i < mMembers.size(); i++) {
            String id = (String) mMembers.elementAt(i);

            if (name.equals(id)) {
                return true;
            }
        }
        return false;
    }

    public void set(String name, Object object) throws EBaseException {
        if (name.equals(ATTR_NAME)) {
            throw new EBaseException(CMS.getUserMessage("CMS_BASE_INVALID_ATTRIBUTE", name));
        } else if (name.equals(ATTR_ID)) {
            throw new EBaseException(CMS.getUserMessage("CMS_BASE_INVALID_ATTRIBUTE", name));
        } else if (name.equals(ATTR_MEMBERS)) {
            mMembers = (Vector) object;
        } else if (name.equals(ATTR_DESCRIPTION)) {
            mDescription = (String) object;
        } else {
            throw new EBaseException(CMS.getUserMessage("CMS_BASE_INVALID_ATTRIBUTE", name));
        }
    }

    public Object get(String name) throws EBaseException {
        if (name.equals(ATTR_NAME)) {
            return getName();
        } else if (name.equals(ATTR_ID)) {
            return getGroupID();
        } else if (name.equals(ATTR_MEMBERS)) {
            return mMembers;
        } else {
            throw new EBaseException(CMS.getUserMessage("CMS_BASE_INVALID_ATTRIBUTE", name));
        }
    }

    public void delete(String name) throws EBaseException {
        throw new EBaseException(CMS.getUserMessage("CMS_BASE_INVALID_ATTRIBUTE", name));
    }

    public Enumeration getElements() {
        return mNames.elements();
    }
}

File: base/tps-tomcat/shared/webapps/tps/js/group.js
/* --- BEGIN COPYRIGHT BLOCK ---
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Copyright (C) 2013 Red Hat, Inc.
 * All rights reserved.
 * --- END COPYRIGHT BLOCK ---
 *
 * @author Endi S. Dewata
 */

var GroupModel = Model.extend({
    urlRoot: "/tps/rest/admin/groups",
    parseResponse: function(response) {
        return {
            id: response.id,
            groupID: response.GroupID,
            description: response.Description
        };
    },
    createRequest: function(attributes) {
        return {
            id: this.id,
            GroupID: attributes.groupID,
            Description: attributes.description
        };
    }
});

var GroupCollection = Collection.extend({
    model: GroupModel,
    urlRoot: "/tps/rest/admin/groups",
    getEntries: function(response) {
        return response.entries;
    },
    getLinks: function(response) {
        return response.Link;
    },
    parseEntry: function(entry) {
        return new GroupModel({
            id: entry.id,
            groupID: entry.GroupID,
            description: entry.Description
        });
    }
});

var GroupMemberModel = Model.extend({
    url: function() {
        var self = this;

        // There's an attribute name mismatch for group ID: the
        // server uses GroupID and the client uses groupID. In other
        // models the mismatch can be translated just fine, but in
        // this model it becomes a problem because the model needs
        // to construct the URL using the attribute.
        //
        // During read operation it needs to use the attribute that's
        // already translated for client (i.e. groupID), but during
        // add it needs to use the attribute meant for server (i.e.
        // GroupID). So the workaround is to read whichever available.
        var groupID = self.get("groupID");        // for read
        groupID = groupID || self.get("GroupID"); // for add

        var url = "/tps/rest/admin/groups/" + groupID + "/members";

        // append member ID for read
        if (self.id) url = url + "/" + self.id;

        return url;
    },
    parseResponse: function(response) {
        return {
            id: response.id,
            memberID: response.id,
            groupID: response.GroupID
        };
    },
    createRequest: function(entry) {
        return {
            id: entry.memberID,
            GroupID: entry.groupID
        };
    }
});

var GroupMemberCollection = Collection.extend({
    initialize: function(models, options) {
        var self = this;
        GroupMemberCollection.__super__.initialize.call(self, models, options);
        options = options || {};
        self.groupID = options.groupID;
        self.urlRoot = "/tps/rest/admin/groups/" + self.groupID + "/members";
    },
    getEntries: function(response) {
        return response.Member;
    },
    getLinks: function(response) {
        return response.Link;
    },
    model: function(attrs, options) {
        return new GroupMemberModel({
            groupID: this.groupID
        });
    },
    parseEntry: function(entry) {
        return new GroupMemberModel({
            id: entry.id,
            memberID: entry.id,
            groupID: entry.GroupID
        });
    }
});

var GroupMembersTableItem = TableItem.extend({
    initialize: function(options) {
        var self = this;
        GroupMembersTableItem.__super__.initialize.call(self, options);
    },
    renderColumn: function(td, templateTD) {
        var self = this;

        GroupMembersTableItem.__super__.renderColumn.call(self, td, templateTD);

        $("a", td).click(function(e) {
            e.preventDefault();
            self.table.open(self);
        });
    }
});

var GroupPage = EntryPage.extend({
    initialize: function(options) {
        var self = this;
        GroupPage.__super__.initialize.call(self, options);
    },
    setup: function() {
        var self = this;

        GroupPage.__super__.setup.call(self);

        var dialog = self.$("#member-dialog");

        var addDialog = new Dialog({
            el: dialog,
            title: "Add Member",
            readonly: ["groupID"],
            actions: ["cancel", "add"]
        });

        var editDialog = new Dialog({
            el: dialog,
            title: "Member",
            readonly: ["groupID", "memberID"],
            actions: ["close"]
        });

        self.membersTable = new ModelTable({
            el: self.$("table[name='members']"),
            pageSize: 10,
            addDialog: addDialog,
            editDialog: editDialog,
            tableItem: GroupMembersTableItem,
            parent: self
        });
    },
    renderContent: function() {
        var self = this;

        GroupPage.__super__.renderContent.call(self);

        // Since the members table is backed by a REST resource any
        // changes will be executed immediately even if the page is
        // in view mode. To avoid confusion, the members table will
        // be disabled in page edit mode.
        if (self.mode == "edit") {
            // In page edit mode, the members tables is read-only.
            self.membersTable.mode = "view";

        } else { // self.mode == "view"
            // In page view mode, the members table is editable.
            self.membersTable.mode = "edit";
        }

        self.membersTable.collection = new GroupMemberCollection(null, { groupID: self.entry.id });
        self.membersTable.render();
    }
});

var GroupsTable = ModelTable.extend({
    initialize: function(options) {
        var self = this;
        GroupsTable.__super__.initialize.call(self, options);
    },
    add: function() {
        var self = this;

        window.location.hash = "#new-group";
    }
});

var GroupsPage = Page.extend({
    load: function() {
        var self = this;

        var table = new GroupsTable({
            el: $("table[name='groups']"),
            collection: new GroupCollection()
        });

        table.render();
    }
});
